pkg/virt-controller/watch/application.go

```diff
@@ -173,31 +188,58 @@ func Execute() {
 
 	app.migrationInformer = app.informerFactory.VirtualMachineInstanceMigration()
 
-	cache.WaitForCacheSync(stopChan, configMapInformer.HasSynced)
-	app.clusterConfig = virtconfig.NewClusterConfig(configMapInformer, app.kubevirtNamespace)
-
-	if app.clusterConfig.DataVolumesEnabled() {
+	if app.hasCdi {
 		app.dataVolumeInformer = app.informerFactory.DataVolume()
-		log.Log.Infof("DataVolume integration enabled")
+		log.Log.Infof("CDI detected, DataVolume integration enabled")
 	} else {
 		// Add a dummy DataVolume informer in the event datavolume support
 		// is disabled. This lets the controller continue to work without
 		// requiring a separate branching code path.
 		app.dataVolumeInformer = app.informerFactory.DummyDataVolume()
-		log.Log.Infof("DataVolume integration disabled")
+		log.Log.Infof("CDI not detected, DataVolume integration disabled")
 	}
 
 	app.initCommon()
 	app.initReplicaSet()
 	app.initVirtualMachines()
 	app.initDisruptionBudgetController()
 	app.initEvacuationController()
-	app.Run()
+	go app.Run()
+
+	select {
+	case <-reInitChan:
+		cancel()
+	}
+}
+
+// Detects if a config has been applied that requires
+// re-initializing virt-controller.
+func (vca *VirtControllerApp) initializationMonitor() <-chan struct{} {
```

Comment:
Can we implement a callback model on the configmap? I fear that these loops will spread all over the place otherwise. (Something like execute callbacks if the config map changed via listwatcher).

Reply:
in this case we care about a callback for crds.  It can be done. I'm unsure if I want to though.

I think virt-controller is the only component that will need such a loop. I had planned to introduce logic here that requires multiple consecutive  loop iterations to detect a change before re-initializing in order to prevent virt-controller from re-initializing unless the detected change really stuck...

So if for example the DataVolume CRD was deleted and immediately re-added, virt-controller wouldn't re-initialize if we have the monitor loop observe something like a "consecutive fail count" before exiting.

Reply:
i went ahead and added a callback based on informers